### PR TITLE
fix(checkbox): Fix indeterminate state not getting correctly applied

### PR DESCRIPTION
**Pull request by vibdev**
## Summary

Fixes: #2280   

The underlying checkbox was not getting the indeterminate attribute applied and we were relying on it being both checked and indeterminate to update the visual display. This change sets the attribute and updates the CSS to also check the `:indeterminate` pseudo class. Indeterminate checkboxes are a bit odd as [the state can't be directly changed via HTML and needs to be set with a ref](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox#indeterminate_state_checkboxes)

## Release Category
Components

---

## Checklist

- [x] MDX documentation adheres to Canvas Kit's [standard MDX template](https://github.com/Workday/canvas-kit/discussions/1131)
- [x] Label `ready for review` has been added to PR

## For the Reviewer

- [x] PR title is short and descriptive
- [x] PR summary describes the change (Fixes/Resolves linked correctly)
- [x] PR Release Notes describes additional information useful to call out in a release message or removed if not applicable
- [x] Breaking Changes provides useful information to upgrade to this code or removed if not applicable

## Areas for Feedback? (optional)

- [ ] Code
- [ ] Documentation
- [x] Testing
- [ ] Codemods

## Thank You Gif (optional)

a Simpsons character holding up a diploma "With minor indetermination"

### CODE REVIEW

File: modules/react/checkbox/lib/Checkbox.tsx [hunk 1]
```diff
@@ -330,18 +339,18 @@ export const Checkbox = createComponent('input')({
...
             {...elemProps}
           />
           <CheckboxRipple variant={variant} />
           <CheckboxBackground variant={variant} checked={checked} disabled={disabled}>
-            <CheckboxCheck checked={checked}>
+            <CheckboxCheck checked={checked} indeterminate={indeterminate}>
               {indeterminate ? (
                 <IndeterminateBox variant={variant} />
-              ) : (
+              ) : checked ? (
```
Comment (vibdev) [inline modules/react/checkbox/lib/Checkbox.tsx:353]: This fixes a different sneaky bug where the checkbox flashes the checked icon when moving between indeterminate true to false

Comment (jamesfan) [reply modules/react/checkbox/lib/Checkbox.tsx:352]: Nice catch!

File: modules/react/checkbox/lib/Checkbox.tsx
Comment (jamesfan) [inline modules/react/checkbox/lib/Checkbox.tsx:1]: Can you look into the regressions in the Chromatic tests? You can disregard the `Tooltips Placements` changes, but we should address the changes in `Checkbox States` and `Checkbox Themed States`. The regressions look specific to the active states:

image

File: modules/react/checkbox/lib/Checkbox.tsx [hunk 2]
```diff
@@ -219,12 +220,12 @@ const CheckboxInput = styled('input')<CheckboxProps & StyledType>(
           variant === 'inverse' ? `1px solid ${colors.soap300}` : `1px solid ${errorColors.inner}`,
         boxShadow: `0 0 0 1px ${errorColors.inner}, 0 0 0 2px ${errorColors.outer}`,
       },
-      '&:not(:checked):not(:disabled):not(:focus):hover, &:not(:checked):not(:disabled):active': {
+      '&:not(:checked):not(:indeterminate):not(:disabled):not(:focus):hover, &:not(:checked):not(indeterminate):not(:disabled):active': {
```
Comment (jamesfan) [inline modules/react/checkbox/lib/Checkbox.tsx:223]: Think I found the issue triggering the visual regressions mentioned above.
```suggestion
      '&:not(:checked):not(:indeterminate):not(:disabled):not(:focus):hover, &:not(:checked):not(:indeterminate):not(:disabled):active': {
```

Comment (vibdev) [reply modules/react/checkbox/lib/Checkbox.tsx:223]: that could do it, nice catch

File: modules/react/checkbox/stories/examples/Indeterminate.tsx [hunk 3]
```diff
@@ -45,8 +45,8 @@ export const Indeterminate = () => {
 
     const anyToppingChecked = newToppings.filter(topping => topping.checked).length > 0;
     const anyToppingUnchecked = newToppings.filter(topping => !topping.checked).length > 0;
-    setPizzaChecked(anyToppingChecked);
     setPizzaIndeterminate(anyToppingChecked && anyToppingUnchecked);
+    setPizzaChecked(anyToppingChecked && !anyToppingUnchecked);
```
Comment (jamesfan) [inline modules/react/checkbox/stories/examples/Indeterminate.tsx:49]: We want the top-level pizza checkbox to only be checked if _all_ toppings are checked, correct? 

```
!anyToppingUnchecked == "not any toppings are unchecked" == "all toppings are checked"
```

```suggestion
    setPizzaChecked(!anyToppingUnchecked);
```

Alternatively, we could just set an `allToppingsChecked` variable for a less confusing way of putting this 😂

Comment (vibdev) [reply modules/react/checkbox/stories/examples/Indeterminate.tsx]: oof yeah the first part is redundant, I like the idea of making a new var to be more clear on the intentions here

File: modules/react/checkbox/lib/Checkbox.tsx [hunk 4]
```diff
@@ -360,6 +368,7 @@ export const Checkbox = createComponent('input')({
             disabled={disabled}
             variant={variant}
             paddingInlineStart={checkboxLabelDistance}
+            cursor="pointer"
```
Comment (vibdev) [inline modules/react/checkbox/lib/Checkbox.tsx:371]: @jamesfan just noticed the label doesn't have the correct cursor even though it is clickable

File: modules/react/checkbox/lib/Checkbox.tsx [hunk 5]
```diff
@@ -322,6 +323,13 @@ export const Checkbox = createComponent('input')({
     Element
   ) => {
     const inputId = useUniqueId(id);
+    const {localRef, elementRef} = useLocalRef(ref);
+    React.useEffect(() => {
+      if (typeof indeterminate === 'boolean' && localRef.current) {
+        localRef.current.indeterminate = indeterminate;
```
Comment (jamesfan) [inline modules/react/checkbox/lib/Checkbox.tsx:329]: @vibdev Last bit: can you add a Cypress test ensuring `indeterminate` is being set properly on the checkbox?

I'm thinking something which loads the Indeterminate story, checks a topping (any of them), and then verifies the `Supreme Pizza Toppings` checkbox has the `indeterminate` property set. Something like this should do the trick:

```
cy.findByLabelText('Supreme Pizza Toppings').should('have.prop', 'indeterminate', true);
```

**Review by jamesfan — APPROVED** [review]
@vibdev Looks good, thanks for catching this issue!

**Pull request merged**